### PR TITLE
Added PurgeTSS as an Alloy command

### PR DESCRIPTION
**Pull request by macCesar**
<p align="center">
	PurgeCSS logo
</p>

# PurgeTSS as an Alloy command

### Now you can run `purgetss` as an Alloy command!!! 🎉

```bash
> alloy purgetss
```

## Key features of PurgeTSS
- Provides more than 10600 [Tailwind-like utility classes](https://tailwindcss.com/) ready to use in your projects.
- Creates a clean `app.tss` file with only the classes used in your project by parsing all your XML files. **Your original classes will be backed up in `_app.tss`**
- You can customize any of the default classes through a simple configuration file, or you can create new *just-in-time* classes with arbitrary values within the `Views`.
- You can easily use Font Awesome, Material Design and Framework7-Icons fonts in `Buttons` and `Labels`.
- Includes an Animation module to apply a 2D Matrix animation or transformation to any element or to an `Array` of elements.
- Includes a simple two-dimensional Grid System to align and distribute elements in your views.

## Running PurgeTSS for the first time
When you run `alloy purgetss` for the first time in your project it will do the following tasks for you.

### 1. Auto-Run Hook
It will add a `hook` in `alloy.jmk` to auto-run `purgetss` everytime you compile your app. **Very useful when using `liveview`.**

### 2. `purgetss` folder
It will create a `purgetss` folder at the root of your project with the following files and folder:

#### `config.js` file
Where you can customize or create new classes with you own spacing, colors, and margin values and more. See [Customization and Configuration Guide](https://github.com/macCesar/purgeTSS/blob/master/docs/configuring-guide.md) for more details.

#### `tailwind.tss` file
Ir contains all Tailwind-like utility classes, including all your custom classes defined in `config.js`.

#### `definitions.css` file
Is a special `css` file that includes ALL the classes from `fonts.tss` and `tailwind.tss` files ( including the official icon fonts libraries ) to be used with [“Intellisencse for CSS class names in HTML”](https://github.com/macCesar/purgeTSS/blob/master/docs/whats-new/v5.2.1.md#new-definitionscss-file) VS Code extension.

#### `fonts` folder
Where you can add any Icon, Serif, Sans-Serif, Cursive, Fantasy or Monospace icons to be used in your app. See [build-fonts command](https://github.com/macCesar/purgeTSS/blob/master/docs/whats-new/v5.1.0.md#instructions-to-recreate-any-of-the-deleted-libraries) for a step-by-step instructions.

## Availble options
These are the different options ( or subcommands ) that you can pass to `purgetss`.
### `build`
If you need to re-generate `tailwind.tss` and `definitions.css` files run:

```bash
> alloy purgetss build
```
Note that `purgetss` checks for any changes made to `config.js` and regenerates them for you.

### `fonts`
It will copy the **official icon fonts files** for **PurgeTSS** in `app/assets/fonts` including the corresponding CommondJS files in `app/lib`:

```bash
> alloy purgetss fonts
```

```bash
> FontAwesome6Brands-Regular.ttf
> FontAwesome6Free-Regular.ttf
> FontAwesome6Free-Solid.ttf
> Framework7-Icons.ttf
> MaterialIcons-Regular.ttf
> MaterialIconsOutlined-Regular.otf
> MaterialIconsRound-Regular.otf
> MaterialIconsSharp-Regular.otf
> MaterialIconsTwoTone-Regular.otf
```
Please check all available classes for [fontawesome.tss, framework7icons.tss and materialdesign.tss](https://github.com/macCesar/purgeTSS#list-of-available-classes)

### `build-fonts`
To create your custom `fonts.tss` file with all the unicode characters ( for icon fonts ) and style rules from the fonts located in `purgetss/fonts`.

```bash
> alloy purgetss build-fonts
```

Example of a `fonts.tss` file:
```css
'.bevan-italic': { font: { fontFamily: 'Bevan-Italic' } }
'.bevan-regular': { font: { fontFamily: 'Bevan-Regular' } }
'.roboto-regular': { font: { fontFamily: 'Roboto-Regular' } }

'.boxicons': { font: { fontFamily: 'boxicons' } }
'.lineicons': { font: { fontFamily: 'LineIcons' } }

// Unicode Characters
// To use your Icon Fonts in Buttons AND Labels each class sets 'text' and 'title' properties

// boxicons.css
'.bxl-meta': { text: '\uef27', title: '\uef27' }
'.bx-lemon': { text: '\uef28', title: '\uef28' }
'.bxs-lemon': { text: '\uef29', title: '\uef29' }
// ...

// lineicons.css
'.lni-500px': { text: '\uea03', title: '\uea03' }
'.lni-add-files': { text: '\uea01', title: '\uea01' }
'.lni-adobe': { text: '\uea06', title: '\uea06' }
// ...
```

For icon font libraries, it will also create a CommondJS module in `app/lib/` with all the unicode characters.

```javascript
// app/lib/purgetss.fonts.js file
const icons = {
  // boxicons
  'bx': {
    '2fa': '\uef27',
    '3dCubeSphere': '\uef28',
    'aB': '\uef29',
    'accessPoint': '\uef2a',
    'accessPointOff': '\uef2b',
    'accessible': '\uef2c',
    ...
  },
  // tabler-icons
  'ti': {
    '2fa': '\ueca0',
    '3dCubeSphere': '\uecd7',
    'aB': '\uec36',
    'accessPoint': '\ued1b',
    'accessPointOff': '\ued1a',
    'accessible': '\ueba9',
    ...
  },
}
```

### alloy purgetss module
To install the `purgetss.ui.js` module in your `lib` folder.

```bash
> alloy purgetss module
```

**PurgeTSS module contains:**
- **Animation**: Apply a 2D Matrix animation or transformation to any element or to an array of elements. **Please check out the new `Animation` module [here](https://github.com/macCesar/purgeTSS/blob/master/docs/whats-new/v2.5.0.md#animation-module).**

### CODE REVIEW

**Comment by m1ga** [comment]
It would be good if it would be a bit more lightweight for people that don't use it. Including it in when installing "just" alloy will install tailwind, fontawesome, framework7-icons. Not everyone uses them but it will still install those dependencies.

**An idea:**

add the command and add a message that you need to install it when it is not found with `npm i -g purgetss`. But make a PR to add it to the front-page: https://titaniumsdk.com/ ("Feature: Use tailwind with purgetss") and add a "how to use it" guide at https://titaniumsdk.com/guide/Alloy_Framework/ 
Ti will have the SEO benefit of "Tailwind" and purgetss is officially feature as a package.

**Comment by caspahouzer** [comment]
We should definitely put that under the guides section of titaniumsdk.com

and let the user choose, which font he wants to install. I prefer fontawesome, Android developer often will choose the material icons.

File: Alloy/alloy.js
Comment (cb1kenobi) [inline Alloy/alloy.js:141]: I for one wouldn't miss this ascii art should it go away.

File: Alloy/commands/purgetss/index.js
Comment (cb1kenobi) [inline Alloy/commands/purgetss/index.js:7]: Do the commands need to be executed in any particular order? Do you need to init before you build?

Comment (macCesar) [reply Alloy/commands/purgetss/index.js:7]: Not at all... It is there.. for people who want to execute several commands at once...

The 'main' command is a one-time execution of `alloy purgetss`.

- From there.. On every compilation it will automatically `parse` all `XML` files y create a new `app.tss` file with only the classes used in `Views`.

- It will automatically re-create the `tailwind.tss` and the `definitions.css` files when there is an update in `config.js`.

Comment (macCesar) [reply Alloy/commands/purgetss/index.js:7]: The `build` command is to re-create `tailwind.tss` without compiling the app... For example to test a new class or a `config.js` change.

The `fonts` command is to copy the icon font libraries... again a one-time command... Unless they deleted the `assets/fonts` folder and want to copy the icon libraries again.

The `build-fonts` command is also a one-time command, unless they added new fonts to `purgetss/fonts` folder and want to re-create the custom `fonts.tss` file with new unicode characters and style rules.

**Comment by macCesar** [comment]
> It would be good if it would be a bit more lightweight for people that don't use it. Including it in when installing "just" alloy will install tailwind, fontawesome, framework7-icons. Not everyone uses them but it will still install those dependencies.

@m1ga ... to summarize:

You want me to add the command without adding the node dependency to Alloy... And to check if `purgetss` is installed ( globally ). If not.. show the instructions to install it.. Right ???

If so.. I like it... It might be a much better solution!! :-)

I'm going to test it!!

**Comment by m1ga** [comment]
> You want me to add the command without adding the node dependency to Alloy... And to check if `purgetss` is installed ( globally ). If not.. show the instructions to install it.. Right ???
> 
> If so.. I like it... It might be a much better solution!! :-)
> 
> I'm going to test it!!

exactly 👍 I think most of users currently don't use it (there are even some that don't use Alloy 😢 ) and we already have too many node dependencies in all projects. The node/alloy folder has 5500 files at the end, purgetss 13.000 😄 think it should be a user decision to add those and not add it automatically by just installing Alloy. 
And you could even add a `not installed - shall I install it for you now` option when you try to run `alloy purgetss` for the first time. The `use sudo npm i -g purgetss to install it` message should be a lot easier to add. Maybe do that for the first version in Alloy and when you see a high demand add the automatic install after the first call option.

**Comment by macCesar** [comment]
Thanks Michale.. I'll see what I can do!!

**Comment by macCesar** [comment]
Michael.. How about this...

If I remove _FontAwesome_ and _framework7_ icons... The total amount of files are around 5,500. files and weighing 55 MB.

**That way `purgetss` will always be installed and ready to use when needed !!! :-D**

Instead of telling the user to install it manually or even automatically when they run `alloy purgetss` for the very first time.

**It is like ALL those [samples files](https://github.com/tidev/alloy/tree/master/samples) that Alloy includes... And no one that I know of has ever used.**

_Sebastian won't like the idea of removing FA.. :-) But I can put instructions on how to re-create them with the `build-fonts` command. ;-)_

**Comment by macCesar** [comment]
> I think most of users currently don't use it

IF we add the instructions in the font-page and in alloy/guides section.. 

**People WILL know about it and start to use it !!** 😄

File: Alloy/alloy.js
Comment (caspahouzer) [reply Alloy/alloy.js:141]: I'm also good with that

**Comment by caspahouzer** [comment]
> _Sebastian won't like the idea of removing FA.. 😟 But I can put instructions on how to re-create both libraries with the `build-fonts` command._ 😉

What I mean is more like
```
> alloy purgetss fonts
Which font do you want to use? FontAwesome (fa), Material Icons (ma), Framework7-Icons (f7), all (a):
```
and then download the chosen one or all

**Comment by caspahouzer** [comment]
> > I think most of users currently don't use it

That's right but could be changed if they see the simplicity of use and how easy they can make awesome designs without looking every 5 minutes in the documentation, how this and that is done 
> 
> IF we add the instructions in the front-page and in alloy/guides section..
> 
> **People WILL know about it and start to use it !!** 😄

I would like a new chapter `PurgeTSS` in the [Alloy Guide](https://titaniumsdk.com/guide/Alloy_Framework/Alloy_Guide/)

As I wrote, I can support you with the guidelines there

**Comment by caspahouzer** [comment]
@m1ga would it be a real problem if purgetss is installed as dependency? I know, it includes a lot of files of fontawesome e.g.

But I've installed packages which are even bigger with more files. And I have no idea, if alle files are used 😄

Do you think it's a space or download problem?

**Comment by m1ga** [comment]
not a real problem but we had discussions before like "why is this a module and not included in the SDK". In my opinion the core should not include extra parts that are not necessary. 

If it's just purgetss without a whole bunch of other dependencies by itself (so the "ask first and then install" way described earlier) then its fine. We already have so many dependencies to update in titanium-sdk and alloy (and the other npm packages). 

I'm still fine with plain alloy and don't use tailwind in webdev either but thats just my preference. But it shouldn't be installed and potentially break my builds when I don't use it or produce more "package is outdated" issues as we already have :smile: Also it would be installed in the github actions.

Again: big fan of featuring it and putting it on the official pages but I personally would prefer to ask the user when running `alloy purgetss`.

**Comment by macCesar** [comment]
I already refactored the command to work with the global version of `purgetss`.. 

I'm testing different options to ask the user to install it when not available... or maybe install it automatically after asking for permissions.. 

After that.. I'm going to publish the PR.. 👍🏼

**Comment by cb1kenobi** [comment]
In general, I love this. I think this is a really great addition. I love that PurgeTSS is just a dependency which a which allows @macCesar to improve it without TiDev needing to release a new Alloy, though we'd be more than happy to do so.

The only thing I caution is the Alloy CLI needs to die. It should have never existed. Alloy should just be a Titanium CLI plugin that provides a new project template and a build hook. The plugin would probably be shipped with the Titanium CLI so no more `npm i -g alloy`.

Titanium CLI plugins can define commands, so adding a `ti purgetss` command would be trivial. Independent of this Alloy plugin thing, we can update the `ti create` command to ask you if you want to create a classic app or an Alloy app, if you want JS or TS, and if Alloy, do you want to use PurgeTSS. In fact, I already created the app templates: https://github.com/tidev/titanium-templates/tree/master/packages.

In the future, we will improve the Titanium CLI. In the meantime, I'll leave it to @caspahouzer and @m1ga to functional test, approve, and merge this PR. Thanks @macCesar!

File: Alloy/alloy.js
Comment (cb1kenobi) [reply Alloy/alloy.js:141]: Leave the banner alone. It's outside the context of this PR. We'll fix it later.

**Review by cb1kenobi — APPROVED** [review]

**Comment by macCesar** [comment]
Thank you Chris for believing in my project and for considering it part of Alloy/Titanium!!!

It means a lot to me!!

**Comment by caspahouzer** [comment]
On first run of `alloy purgetss` add the following files to the `.gitignore` if it exists

- purgetss/fontawesome.tss
- purgetss/tailwind.tss
- purgetss/fonts.tss
- purgetss/definitions.css
- app.tss

Since these files are generated on build and rebuild with liveview, there is no need to have them in the users repo

File: Alloy/commands/purgetss/index.js
Comment (m1ga) [inline Alloy/commands/purgetss/index.js:5]: remove comments (and in the line below)

File: Alloy/alloy.js
Comment (m1ga) [inline Alloy/alloy.js:59]: I wonder if we should add a `purgetss` prefix here since it might be useful to use `--debug` for normal alloy commands at one point. I see people use it for plain alloy and not seeing any debug output (like --log-level debug).

**Comment by m1ga** [comment]
@macCesar great work! Just left some comments but I like the new way

File: Alloy/alloy.js
Comment (drauggres) [inline Alloy/alloy.js:61]: You should add purgetss specific options to `program.command('purgetss')`, e.g.
```javascript

program.command('purgetss')
  .option('--debug', 'Show time taken to execute each process.')
```

Also `--all` overlaps with global `.option('-w, --all', 'require flag for generate styles')`

File: Alloy/commands/purgetss/index.js
Comment (drauggres) [inline Alloy/commands/purgetss/index.js:2]: `commandExistsSync` is unused

File: Alloy/commands/purgetss/index.js
Comment (drauggres) [inline Alloy/commands/purgetss/index.js:41]: You can use `die()` from utils here:
https://github.com/tidev/alloy/blob/5e9e25e58ce0fac3e8c12e7583297a5f64d27101/Alloy/utils.js#L521

File: package.json
Comment (drauggres) [inline package.json:42]: `command-exists` is unused

File: package.json
Comment (drauggres) [inline package.json:15]: It is usually not a good idea to increase version in PR.
Maintainers will do it on their own.

**Review by drauggres — CHANGES_REQUESTED** [review]

File: Alloy/commands/purgetss/index.js
Comment (macCesar) [reply Alloy/commands/purgetss/index.js:2]: I forgot to remove `command-exists` didn't work as expected!!

**Review by drauggres — APPROVED** [review]
LGTM

**Comment by caspahouzer** [comment]
I would like to test it in a few hours and will merge then. 

@drauggres @macCesar  should we raise the version?

**Comment by drauggres** [comment]
> should we raise the version?

It's not for me to decide (I was just passing by), but I think we shouldn't do this in feature/bugfix PRs.

Also, I think it would be good to squash all commits (except version changing) into one before merging.

**Comment by caspahouzer** [comment]
@cb1kenobi what do you think about versioning? How to handle that the best way?

**Comment by cb1kenobi** [comment]
@drauggres @caspahouzer dropping in PurgeTSS is a new feature. This could go out in a minor release (e.g. 1.18.0). With that said, one of the initiatives we have right now is dropping Node 12 and older and that affects Alloy too. Dropping Node 12 and older is a breaking change and thus there will be an Alloy v2. I know when we rev a major we think of cool new features like PurgeTSS, but I can't say when v2 will ship. Could be a couple weeks, could be a couple months. Dropping Node 12 is high on my list, so it won't be forever.

**Comment by cb1kenobi** [comment]
@caspahouzer yup, bump please.

**Comment by macCesar** [comment]
@caspahouzer do you want me to do something else ?? Or is it up to the maintainers now ??

**Comment by cb1kenobi** [comment]
@macCesar at this point, you need the merge conflicts resolved, bump the version, and most importantly, someone to actually pull and TEST.

**Comment by caspahouzer** [comment]
I'll test tomorrow and comment the results. 

If all good, I'll approve and merge

**Comment by caspahouzer** [comment]
> @caspahouzer do you want me to do something else ?? Or is it up to the maintainers now ??

Nope. Let you know my test results. Maybe I'm chat to you on slack for any questions

**Review by caspahouzer — CHANGES_REQUESTED** [review]
Revert that changes please in favor of https://github.com/tidev/alloy/pull/1326

**Comment by cb1kenobi** [comment]
Can someone please pull this PR and test it so we can merge it?

File: Alloy/alloy.js
Comment (caspahouzer) [inline Alloy/alloy.js:137]: No conflicts with https://github.com/tidev/alloy/pull/1326

**Comment by caspahouzer** [comment]
I'm happy with all changes and all my test PASS

**Review by caspahouzer — APPROVED** [review]

**Comment by cb1kenobi** [comment]
Yay! :clap:

**Pull request merged**
